core-primitives/stf-state-handler/src/state_snapshot_repository.rs

```diff
@@ -0,0 +1,445 @@
+				// in general was successful, just a side-effect that failed.
+				error!("Failed to remove state, with id '{}': {:?}", snapshot_metadata.state_id, e);
+			}
+		}
+	}
+
+	fn write_new_state(
+		&self,
+		shard_identifier: &ShardIdentifier,
+		state: State,
+	) -> Result<(HashType, StateId)> {
+		let state_id = generate_current_timestamp_state_id();
+		let state_hash = self.file_io.write(shard_identifier, state_id, state)?;
+		Ok((state_hash, state_id))
+	}
+
+	fn load_state(
+		&self,
+		shard_identifier: &ShardIdentifier,
+		snapshot_metadata: &StateSnapshotMetaData<HashType>,
+	) -> Result<State> {
+		self.file_io.load(shard_identifier, snapshot_metadata.state_id)
+	}
+}
+
+impl<FileIo, State, HashType> VersionedStateAccess
+	for StateSnapshotRepository<FileIo, State, HashType>
+where
+	FileIo: StateFileIo<StateType = State, HashType = HashType>,
+	HashType: Copy + Eq + Debug,
+{
+	type StateType = State;
+	type HashType = HashType;
+
+	fn load_latest(&self, shard_identifier: &ShardIdentifier) -> Result<Self::StateType> {
+		let latest_snapshot_metadata = self.get_latest_snapshot_metadata(shard_identifier)?;
+		self.file_io.load(shard_identifier, latest_snapshot_metadata.state_id)
+	}
+
+	fn update(
+		&mut self,
+		shard_identifier: &ShardIdentifier,
+		state: Self::StateType,
+	) -> Result<Self::HashType> {
+		if !self.shard_exists(shard_identifier) {
+			return Err(Error::InvalidShard(*shard_identifier))
+		}
+
+		let (state_hash, state_id) = self.write_new_state(shard_identifier, state)?;
+		let cache_size = self.snapshot_history_cache_size;
+
+		let snapshot_history = self.get_snapshot_history_mut(shard_identifier)?;
+		snapshot_history.push_front(StateSnapshotMetaData::new(state_hash, state_id));
+
+		// In case we're above max queue size we remove the oldest entries and corresponding files
+		if snapshot_history.len() > cache_size {
+			self.prune_snapshot_history_by_range(shard_identifier, cache_size..)?;
+		}
+
+		Ok(state_hash)
+	}
+
+	fn revert_to(
+		&mut self,
+		shard_identifier: &ShardIdentifier,
+		state_hash: &Self::HashType,
+	) -> Result<Self::StateType> {
+		let snapshot_history = self.get_snapshot_history(shard_identifier)?;
+
+		// We use `position()` instead of `find()`, because it then allows us to easily drain
+		// all the newer states.
+		let snapshot_metadata_index = snapshot_history
+			.iter()
+			.position(|fmd| fmd.state_hash == *state_hash)
+			.ok_or_else(|| Error::StateNotFoundInRepository(format!("{:?}", state_hash)))?;
+
+		// Should never fail, since we got the index from above, with `position()`.
+		let snapshot_metadata = snapshot_history
+			.get(snapshot_metadata_index)
+			.ok_or_else(|| Error::StateNotFoundInRepository(format!("{:?}", state_hash)))?;
+
+		let state = self.load_state(shard_identifier, snapshot_metadata)?;
+
+		// Remove any state versions newer than the one we're resetting to
+		// (do this irreversible operation last, to ensure the loading has succeeded)
```

Comment:
Yes, and in case we have an error and don't succeed in returning the loaded state, we don't want to have pruned everything already. So the pruning comes last, once we know everything else succeeded.

Reply:
Ah ok, thanks for clearing this up, I misread the comment. 👍